private/tools/java/com/github/bazelbuild/rules_jvm_external/javadoc/JavadocJarMaker.java

```diff
@@ -248,16 +276,81 @@ private static void readSourceFiles(
             ByteStreams.copy(zis, out);
           }
 
-          fileManager.getJavaFileObjects(target.toFile()).forEach(sources::add);
+          fileManager
+              .getJavaFileObjects(target.toFile())
+              .forEach(
+                  s -> {
+                    String p = extractPackageName(s);
+                    sources.computeIfAbsent(p, k -> new ArrayList<>()).add(s);
+                  });
+        }
+      }
+    }
+  }
 
-          String[] segments = name.split("/");
-          if (segments.length > 0 && !"META-INF".equals(segments[0])) {
-            topLevelPackages.add(segments[0]);
+  // If the package ends in .* , then look for all subpackages in packages set
+  private static Set<String> expandExcludedPackages(
+      Set<String> excludedPackages, Set<String> packages) {
+    Set<String> expandedPackages = new HashSet<>();
+
+    for (String excludedPackage : excludedPackages) {
+      if (excludedPackage.endsWith(".*")) {
+        String basePackage = excludedPackage.substring(0, excludedPackage.length() - 2);
+        for (String pkg : packages) {
+          if (pkg.startsWith(basePackage)) {
+            expandedPackages.add(pkg);
           }
+        }
+      } else {
+        expandedPackages.add(excludedPackage);
+      }
+    }
+
+    return expandedPackages;
+  }
 
-          fileNames.add(name);
+  // Extract the package name from the contents of the file
+  private static String extractPackageName(JavaFileObject fileObject) {
+    try (Reader reader = fileObject.openReader(true);
+        BufferedReader bufferedReader = new BufferedReader(reader)) {
+      String line;
+      while ((line = bufferedReader.readLine()) != null) {
+        if (line.startsWith("package ")) {
+          return line.substring("package ".length(), line.indexOf(';')).trim();
+        }
+
+        // Stop looking if we hit the class or interface declaration
+        if (line.startsWith("public")
+            || line.startsWith("class")
+            || line.startsWith("interface")
+            || line.startsWith("enum")) {
+          break;
```

Comment:
optionally: Set.of("public", "class", "interface", "enum").contains(line)
(can also move the set outside)

Reply:
Went with `keywords.stream().anyMatch(line::startsWith)`